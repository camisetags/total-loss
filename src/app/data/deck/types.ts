export interface Deck {
  id: number;
  selected: boolean;
  name: string;
  description: string;
  cards: object[];
}

export type DeckState = {
  [id: number]: Deck;
};
