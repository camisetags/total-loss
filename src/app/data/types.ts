import { Actions as DeckActions } from './deck/actionTypes';
import { DeckState } from './deck/types';

export type RootAction = DeckActions[keyof DeckActions];

export type RootState = {
  deck: DeckState;
};
