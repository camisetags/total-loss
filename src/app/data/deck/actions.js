import * as deckTypes from './actionTypes';
import deckApi from './api';

export const getDeckList = () => async (dispatch) => {
  const response = await deckApi.getList();
  dispatch({
    type: deckTypes.LIST_DECKS,
    data: response.data.results,
  });
};

export const selectDeck = deckId => async (dispatch) => {
  const response = await deckApi.getById(deckId);
  dispatch({
    type: deckTypes.SELECT_DECK,
    data: {
      name: response.data.name,
      description: response.data.description,
      cards: [...response.data.cards],
    },
  });
};

export default getDeckList;
