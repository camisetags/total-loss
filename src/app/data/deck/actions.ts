import { normalize } from 'normalizr';
import { Deck } from './types';
import { deckSchema } from './schemas';
import { LIST_DECKS, SELECT_DECK, Actions as DeckActions } from './actionTypes';
import deckApi from './api';

import { Dispatch } from 'redux';

export const getDeckList = () => async (dispatch: Dispatch<DeckActions>) => {
  const response = await deckApi.getList();
  const deckList = response.data.results.map((deck: Deck) => ({
    ...deck,
    selected: false,
  }));

  dispatch({
    type: LIST_DECKS,
    data: normalize(deckList, [deckSchema]).entities.decks,
  });
};

export const selectDeck = (deckId: number) => async (dispatch: Dispatch<DeckActions>) => {
  const response = await deckApi.getById(deckId);
  const deck = {
    id: deckId,
    selected: true,
    name: response.data.name,
    description: response.data.description,
    cards: [...response.data.cards],
  };

  dispatch({
    type: SELECT_DECK,
    data: normalize(deck, deckSchema).entities.decks,
  });
};

export default getDeckList;
