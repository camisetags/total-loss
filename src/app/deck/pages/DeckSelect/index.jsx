/* eslint no-console: 0 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { getDeckList } from '../../actions';
import DeckSelectComponent from './component';

class DeckSelect extends React.Component {
  static get propTypes() {
    return {
      fetchDecks: PropTypes.func.isRequired,
      decks: PropTypes.arrayOf(PropTypes.object).isRequired,
    };
  }

  componentDidMount() {
    this.props.fetchDecks();
  }

  render() {
    return <DeckSelectComponent {...this.props} />;
  }
}

const mapStateToProps = state => ({
  decks: state.deck.list,
});

const mapDispatchToProps = dispatch => ({
  fetchDecks: () => {
    dispatch(getDeckList());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(DeckSelect);
