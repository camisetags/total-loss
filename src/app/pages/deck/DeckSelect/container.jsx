import React, { Component } from 'react';
import { connect } from 'react-redux';
import { func, arrayOf, object } from 'prop-types';

import { objToArray } from '../../../helpers/utils';
import { getDeckList, selectDeck } from '../../../data/deck/actions';

import DeckSelectComponent from './component';
import Loading from '../../../components/Loading';

export class DeckSelect extends Component {
  static propTypes = {
    fetchDecks: func.isRequired,
    fetchDeck: func.isRequired,
    decks: arrayOf(object).isRequired,
  };

  componentDidMount() {
    this.props.fetchDecks();
  }

  selectDeck = (deckId) => {
    this.props.fetchDeck(deckId);
  };

  render() {
    if (this.props.decks.length === 1) {
      return <Loading />;
    }

    return <DeckSelectComponent decks={this.props.decks} selectDeck={this.selectDeck} />;
  }
}

const mapStateToProps = state => ({
  decks: objToArray(state.deck),
});

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchDecks: () => dispatch(getDeckList()),
  fetchDeck: deckId => dispatch(selectDeck(deckId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(DeckSelect);
