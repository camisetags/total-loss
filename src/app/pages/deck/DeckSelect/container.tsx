import * as React from 'react';
import { connect, Dispatch } from 'react-redux';

import { objToArray } from '../../../helpers/utils';
import { getDeckList, selectDeck } from '../../../data/deck/actions';
import { Deck } from '../../../data/deck/types';
import { RootState } from '../../../data/types';

import DeckSelectComponent from './component';
import Loading from '../../../components/Loading';

type DeckSelectProps = {
  fetchDecks: Function;
  fetchDeck: Function;
  decks: Deck[];
};

export class DeckSelect extends React.Component<DeckSelectProps, {}> {
  componentDidMount() {
    this.props.fetchDecks();
  }

  selectDeck(deckId: number) {
    this.props.fetchDeck(deckId);
  }

  render() {
    if (this.props.decks.length === 1) {
      return <Loading />;
    }

    return (
      <DeckSelectComponent
        decks={this.props.decks}
        selectDeck={(deckId: number) => this.selectDeck(deckId)}
      />
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  decks: Object.values(state.deck),
});

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  fetchDecks: () => dispatch(getDeckList()),
  fetchDeck: (deckId: number) => dispatch(selectDeck(deckId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(DeckSelect as any);
