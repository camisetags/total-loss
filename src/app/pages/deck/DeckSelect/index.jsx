import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { getDeckList, selectDeck } from '../../../data/deck/actions';

import DeckSelectComponent from './component';
import Loading from '../../../components/Loading';

class DeckSelect extends React.Component {
  static get propTypes() {
    return {
      fetchDecks: PropTypes.func.isRequired,
      fetchDeck: PropTypes.func.isRequired,
      decks: PropTypes.arrayOf(PropTypes.object).isRequired,
      deck: PropTypes.object,
    };
  }

  componentDidMount() {
    this.props.fetchDecks();
  }

  selectDeck(deckId) {
    this.props.fetchDeck(deckId);
  }

  render() {
    if (this.props.decks.length === 1) {
      return <Loading />;
    }

    return <DeckSelectComponent {...this.props} selectDeck={deckId => this.selectDeck(deckId)} />;
  }
}

const mapStateToProps = state => ({
  decks: state.deck.list,
});

const mapDispatchToProps = dispatch => ({
  fetchDecks: () => dispatch(getDeckList()),
  fetchDeck: deckId => dispatch(selectDeck(deckId)),
});

DeckSelect.defaultProps = {
  deck: {},
};

export default connect(mapStateToProps, mapDispatchToProps)(DeckSelect);
