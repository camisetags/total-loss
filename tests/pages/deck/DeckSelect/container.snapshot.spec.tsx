import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter, Link } from 'react-router-dom';

import { DeckSelect as Container } from '../../../../src/app/pages/deck/DeckSelect/container';
import { Deck, DeckState } from '../../../../src/app/data/deck/types';

describe('DeckSelect Container', () => {
  const deckInitialState = {
    1: {
      id: 1,
      name: 'Test deck 1',
      description: 'Description deck one',
    },
    2: {
      id: 2,
      name: 'Test deck 2',
      description: 'Description deck on 2',
    },
    3: {
      id: 3,
      name: 'Test deck 3',
      description: 'Description deck 3',
    },
  };

  const mockFetchDecks = jest.fn();
  const mockFetchDeck = jest.fn();
  const componentDidMountSpy = jest.spyOn(Container.prototype, 'componentDidMount');

  beforeEach(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  afterEach(() => {
    componentDidMountSpy.mockClear();
  });

  it('should render loading component', () => {
    const wrapper = Enzyme.mount(
      <Container
        decks={[deckInitialState[0] as Deck]}
        fetchDeck={() => mockFetchDeck()}
        fetchDecks={() => mockFetchDecks()}
      />,
    );

    const loading = wrapper.find('.loading');
    expect(loading.text()).toBe(' Carregando... ');
  });

  it('should render deckSelect page component inside container', () => {
    const rendered = renderer.create(
      <MemoryRouter>
        <Container
          decks={(Object as any).values(deckInitialState) as Deck[]}
          fetchDeck={() => mockFetchDeck()}
          fetchDecks={() => mockFetchDecks()}
        />
      </MemoryRouter>,
    );

    expect(rendered.toJSON).toMatchSnapshot();
  });

  it('should call fetch decks on componentDidMount', () => {
    const wrapper = Enzyme.mount(
      <Container
        decks={[deckInitialState[0]]}
        fetchDeck={() => mockFetchDeck()}
        fetchDecks={() => mockFetchDecks()}
      />,
    );

    wrapper.instance().componentDidMount();

    expect(componentDidMountSpy).toHaveBeenCalled();
    expect(mockFetchDecks.mock.calls.length).toBeGreaterThanOrEqual(1);
  });

  it('should call fetch single deck', () => {
    const wrapper = Enzyme.mount(
      <MemoryRouter>
        <Container
          decks={(Object as any).values(deckInitialState) as Deck[]}
          fetchDeck={() => mockFetchDeck()}
          fetchDecks={() => mockFetchDecks()}
        />
      </MemoryRouter>,
    );

    wrapper
      .find(Link)
      .at(0)
      .simulate('click');

    expect(mockFetchDeck.mock.calls.length).toEqual(1);
  });
});
