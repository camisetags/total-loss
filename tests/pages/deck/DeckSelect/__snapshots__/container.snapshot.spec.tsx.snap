// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckSelect Container should render deckSelect page component inside container 1`] = `[Function]`;
